ICE: failed to get layout

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
use std::mem;

pub struct S<T: Tr> {
    pub f: <T as Tr>::I,
}

pub trait Tr {
   type I: Tr;
}

impl<T: Tr> Tr for S<T> {
    type I = S<S<T>>;
}

impl Tr for () {
    type I = ();
}

fn foo<T: Tr>() -> usize {
    mem::size_of::<S<T>>()
}

fn main() {
    println!("{}", foo::<S<()>>());
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
<version>
```

### Error output

```
<output>
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error[E0391]: cycle detected when computing layout of `S<S<()>>`
  |
  = note: ...which requires computing layout of `<S<()> as Tr>::I`...
  = note: ...which again requires computing layout of `S<S<()>>`, completing the cycle

error: internal compiler error: compiler/rustc_codegen_llvm/src/context.rs:974:13: failed to get layout for `S<S<()>>`: a cycle occurred during layout computation

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/compiler/rustc_errors/src/lib.rs:994:33
stack backtrace:
   0:     0x7f4c65369441 - std::backtrace_rs::backtrace::libunwind::trace::hb9be772067563b53
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f4c65369441 - std::backtrace_rs::backtrace::trace_unsynchronized::h031da30dc183647f
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f4c65369441 - std::sys_common::backtrace::_print_fmt::h71eced90df8827aa
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f4c65369441 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hae72971f9b1c0375
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f4c653ca0df - core::fmt::rt::Argument::fmt::h8155c503d36c9b86
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/core/src/fmt/rt.rs:138:9
   5:     0x7f4c653ca0df - core::fmt::write::hbbb4251c2513c3ef
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/core/src/fmt/mod.rs:1094:21
   6:     0x7f4c6535c591 - std::io::Write::write_fmt::hdcff7ff4dad8c59b
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/io/mod.rs:1712:15
   7:     0x7f4c65369255 - std::sys_common::backtrace::_print::hb8bd80d5dffe8016
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f4c65369255 - std::sys_common::backtrace::print::hd57b6e007b7d9859
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f4c6536bee7 - std::panicking::default_hook::{{closure}}::h92b5b1db37900f49
  10:     0x7f4c6536bcd4 - std::panicking::default_hook::h33b32947a165ebfe
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/panicking.rs:288:9
  11:     0x7f4c685707b5 - <rustc_driver_impl[2c3cd2b22615ac74]::DEFAULT_HOOK::{closure#0}::{closure#0} as core[65ac48eb02ee08bd]::ops::function::FnOnce<(&core[65ac48eb02ee08bd]::panic::panic_info::PanicInfo,)>>::call_once::{shim:vtable#0}
  12:     0x7f4c6536c607 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h65038aad28115701
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/alloc/src/boxed.rs:1999:9
  13:     0x7f4c6536c607 - std::panicking::rust_panic_with_hook::h6818a1a7c65e9cdd
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/panicking.rs:695:13
  14:     0x7f4c68402f91 - std[7223d5cfbaf62610]::panicking::begin_panic::<rustc_errors[eb6afbb30a1cd2c3]::ExplicitBug>::{closure#0}
  15:     0x7f4c68401c26 - std[7223d5cfbaf62610]::sys_common::backtrace::__rust_end_short_backtrace::<std[7223d5cfbaf62610]::panicking::begin_panic<rustc_errors[eb6afbb30a1cd2c3]::ExplicitBug>::{closure#0}, !>
  16:     0x7f4c68401ac6 - std[7223d5cfbaf62610]::panicking::begin_panic::<rustc_errors[eb6afbb30a1cd2c3]::ExplicitBug>
  17:     0x7f4c683f6802 - <rustc_errors[eb6afbb30a1cd2c3]::HandlerInner>::span_bug::<rustc_span[418048bff763fbc6]::span_encoding::Span, alloc[1303c5f908d1c3f4]::string::String>
  18:     0x7f4c683f5f46 - <rustc_errors[eb6afbb30a1cd2c3]::Handler>::span_bug::<rustc_span[418048bff763fbc6]::span_encoding::Span, alloc[1303c5f908d1c3f4]::string::String>
  19:     0x7f4c683faedd - rustc_middle[faa0e51d0d5710e5]::util::bug::opt_span_bug_fmt::<rustc_span[418048bff763fbc6]::span_encoding::Span>::{closure#0}
  20:     0x7f4c683faf0a - rustc_middle[faa0e51d0d5710e5]::ty::context::tls::with_opt::<rustc_middle[faa0e51d0d5710e5]::util::bug::opt_span_bug_fmt<rustc_span[418048bff763fbc6]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f4c683faaba - rustc_middle[faa0e51d0d5710e5]::ty::context::tls::with_context_opt::<rustc_middle[faa0e51d0d5710e5]::ty::context::tls::with_opt<rustc_middle[faa0e51d0d5710e5]::util::bug::opt_span_bug_fmt<rustc_span[418048bff763fbc6]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f4c66d96d44 - rustc_middle[faa0e51d0d5710e5]::util::bug::span_bug_fmt::<rustc_span[418048bff763fbc6]::span_encoding::Span>
  23:     0x7f4c6840d400 - <rustc_codegen_llvm[f6e646cf3297fbe4]::context::CodegenCx as rustc_middle[faa0e51d0d5710e5]::ty::layout::LayoutOfHelpers>::handle_layout_err
  24:     0x7f4c6840d312 - <rustc_codegen_llvm[f6e646cf3297fbe4]::context::CodegenCx as rustc_middle[faa0e51d0d5710e5]::ty::layout::LayoutOf>::spanned_layout_of::{closure#0}
  25:     0x7f4c66db0a4e - <rustc_codegen_llvm[f6e646cf3297fbe4]::context::CodegenCx as rustc_middle[faa0e51d0d5710e5]::ty::layout::LayoutOf>::spanned_layout_of
  26:     0x7f4c66da905f - <rustc_codegen_ssa[ec866eaa3e7fa1d8]::mir::FunctionCx<rustc_codegen_llvm[f6e646cf3297fbe4]::builder::Builder>>::codegen_rvalue_operand
  27:     0x7f4c66d756bc - rustc_codegen_ssa[ec866eaa3e7fa1d8]::mir::codegen_mir::<rustc_codegen_llvm[f6e646cf3297fbe4]::builder::Builder>
  28:     0x7f4c67a12312 - rustc_codegen_llvm[f6e646cf3297fbe4]::base::compile_codegen_unit::module_codegen
  29:     0x7f4c67a11160 - rustc_codegen_llvm[f6e646cf3297fbe4]::base::compile_codegen_unit
  30:     0x7f4c67c55486 - rustc_codegen_ssa[ec866eaa3e7fa1d8]::base::codegen_crate::<rustc_codegen_llvm[f6e646cf3297fbe4]::LlvmCodegenBackend>
  31:     0x7f4c67c54d52 - <rustc_codegen_llvm[f6e646cf3297fbe4]::LlvmCodegenBackend as rustc_codegen_ssa[ec866eaa3e7fa1d8]::traits::backend::CodegenBackend>::codegen_crate
  32:     0x7f4c67803e90 - rustc_interface[145311d2418df64f]::passes::start_codegen
  33:     0x7f4c677fef07 - <rustc_middle[faa0e51d0d5710e5]::ty::context::GlobalCtxt>::enter::<<rustc_interface[145311d2418df64f]::queries::Queries>::ongoing_codegen::{closure#0}::{closure#0}, core[65ac48eb02ee08bd]::result::Result<alloc[1303c5f908d1c3f4]::boxed::Box<dyn core[65ac48eb02ee08bd]::any::Any>, rustc_span[418048bff763fbc6]::ErrorGuaranteed>>
  34:     0x7f4c677fe4e8 - <rustc_interface[145311d2418df64f]::queries::Queries>::ongoing_codegen
  35:     0x7f4c677fdc23 - <rustc_interface[145311d2418df64f]::interface::Compiler>::enter::<rustc_driver_impl[2c3cd2b22615ac74]::run_compiler::{closure#1}::{closure#2}, core[65ac48eb02ee08bd]::result::Result<core[65ac48eb02ee08bd]::option::Option<rustc_interface[145311d2418df64f]::queries::Linker>, rustc_span[418048bff763fbc6]::ErrorGuaranteed>>
  36:     0x7f4c677fb38f - std[7223d5cfbaf62610]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[145311d2418df64f]::util::run_in_thread_pool_with_globals<rustc_interface[145311d2418df64f]::interface::run_compiler<core[65ac48eb02ee08bd]::result::Result<(), rustc_span[418048bff763fbc6]::ErrorGuaranteed>, rustc_driver_impl[2c3cd2b22615ac74]::run_compiler::{closure#1}>::{closure#0}, core[65ac48eb02ee08bd]::result::Result<(), rustc_span[418048bff763fbc6]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[65ac48eb02ee08bd]::result::Result<(), rustc_span[418048bff763fbc6]::ErrorGuaranteed>>
  37:     0x7f4c67e59e9e - <<std[7223d5cfbaf62610]::thread::Builder>::spawn_unchecked_<rustc_interface[145311d2418df64f]::util::run_in_thread_pool_with_globals<rustc_interface[145311d2418df64f]::interface::run_compiler<core[65ac48eb02ee08bd]::result::Result<(), rustc_span[418048bff763fbc6]::ErrorGuaranteed>, rustc_driver_impl[2c3cd2b22615ac74]::run_compiler::{closure#1}>::{closure#0}, core[65ac48eb02ee08bd]::result::Result<(), rustc_span[418048bff763fbc6]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[65ac48eb02ee08bd]::result::Result<(), rustc_span[418048bff763fbc6]::ErrorGuaranteed>>::{closure#1} as core[65ac48eb02ee08bd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x7f4c65376ad5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::haa143886c5139467
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/alloc/src/boxed.rs:1985:9
  39:     0x7f4c65376ad5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha0b7575cc93abe25
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/alloc/src/boxed.rs:1985:9
  40:     0x7f4c65376ad5 - std::sys::unix::thread::Thread::new::thread_start::ha9600bcb75fadfd8
                               at /rustc/129195f57c02c8a8e4cfc4b766d4ff7a3a837882/library/std/src/sys/unix/thread.rs:108:17
  41:     0x7f4c65112bb5 - <unknown>
  42:     0x7f4c65194d90 - <unknown>
  43:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.71.0-nightly (129195f57 2023-05-04) running on x86_64-unknown-linux-gnu

query stack during panic:
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0391`.
```

</p>
</details>


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"atsuzaki"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_codegen_cranelift/src/common.rs
+++ b/compiler/rustc_codegen_cranelift/src/common.rs
@@ -480,7 +480,7 @@ impl<'tcx> LayoutOfHelpers<'tcx> for RevealAllLayoutCx<'tcx> {
         if let LayoutError::SizeOverflow(_) | LayoutError::ReferencesError(_) = err {
             self.0.sess.span_fatal(span, err.to_string())
         } else {
-            span_bug!(span, "failed to get layout for `{}`: {}", ty, err)
+            self.0.sess.span_fatal(span, format!("failed to get layout for `{}`: {}", ty, err))
         }
     }
 }

--- a/compiler/rustc_codegen_gcc/src/context.rs
+++ b/compiler/rustc_codegen_gcc/src/context.rs
@@ -7,6 +7,7 @@ use rustc_codegen_ssa::traits::{
     BaseTypeMethods,
     MiscMethods,
 };
+use rustc_codegen_ssa::errors as ssa_errors;
 use rustc_data_structures::base_n;
 use rustc_data_structures::fx::{FxHashMap, FxHashSet};
 use rustc_middle::span_bug;
@@ -479,7 +480,7 @@ impl<'gcc, 'tcx> LayoutOfHelpers<'tcx> for CodegenCx<'gcc, 'tcx> {
         if let LayoutError::SizeOverflow(_) | LayoutError::ReferencesError(_) = err {
             self.sess().emit_fatal(respan(span, err.into_diagnostic()))
         } else {
-            span_bug!(span, "failed to get layout for `{}`: {}", ty, err)
+            self.tcx.sess.emit_fatal(ssa_errors::FailedToGetLayout { span, ty, err })
         }
     }
 }

--- a/compiler/rustc_codegen_llvm/src/context.rs
+++ b/compiler/rustc_codegen_llvm/src/context.rs
@@ -10,6 +10,7 @@ use crate::value::Value;
 
 use cstr::cstr;
 use rustc_codegen_ssa::base::{wants_msvc_seh, wants_wasm_eh};
+use rustc_codegen_ssa::errors as ssa_errors;
 use rustc_codegen_ssa::traits::*;
 use rustc_data_structures::base_n;
 use rustc_data_structures::fx::FxHashMap;
@@ -1000,7 +1001,7 @@ impl<'tcx> LayoutOfHelpers<'tcx> for CodegenCx<'_, 'tcx> {
         if let LayoutError::SizeOverflow(_) | LayoutError::ReferencesError(_) = err {
             self.sess().emit_fatal(Spanned { span, node: err.into_diagnostic() })
         } else {
-            span_bug!(span, "failed to get layout for `{ty}`: {err:?}")
+            self.tcx.sess.emit_fatal(ssa_errors::FailedToGetLayout { span, ty, err })
         }
     }
 }

--- a/compiler/rustc_codegen_ssa/messages.ftl
+++ b/compiler/rustc_codegen_ssa/messages.ftl
@@ -35,6 +35,8 @@ codegen_ssa_extract_bundled_libs_parse_archive = failed to parse archive '{$rlib
 codegen_ssa_extract_bundled_libs_read_entry = failed to read entry '{$rlib}': {$error}
 codegen_ssa_extract_bundled_libs_write_file = failed to write file '{$rlib}': {$error}
 
+codegen_ssa_failed_to_get_layout = failed to get layout for {$ty}: {$err}
+
 codegen_ssa_failed_to_write = failed to write {$path}: {$error}
 
 codegen_ssa_ignoring_emit_path = ignoring emit path because multiple .{$extension} files were produced

--- a/compiler/rustc_codegen_ssa/src/errors.rs
+++ b/compiler/rustc_codegen_ssa/src/errors.rs
@@ -7,6 +7,7 @@ use rustc_errors::{
     IntoDiagnosticArg,
 };
 use rustc_macros::Diagnostic;
+use rustc_middle::ty::layout::LayoutError;
 use rustc_middle::ty::Ty;
 use rustc_span::{Span, Symbol};
 use rustc_type_ir::FloatTy;
@@ -1028,6 +1029,15 @@ pub struct TargetFeatureSafeTrait {
     pub span: Span,
     #[label(codegen_ssa_label_def)]
     pub def: Span,
+}
+
+#[derive(Diagnostic)]
+#[diag(codegen_ssa_failed_to_get_layout)]
+pub struct FailedToGetLayout<'tcx> {
+    #[primary_span]
+    pub span: Span,
+    pub ty: Ty<'tcx>,
+    pub err: LayoutError<'tcx>,
 }
 
 #[derive(Diagnostic)]

--- a/compiler/rustc_middle/src/ty/layout.rs
+++ b/compiler/rustc_middle/src/ty/layout.rs
@@ -4,7 +4,9 @@ use crate::query::TyCtxtAt;
 use crate::ty::normalize_erasing_regions::NormalizationError;
 use crate::ty::{self, ConstKind, ReprOptions, Ty, TyCtxt, TypeVisitableExt};
 use rustc_error_messages::DiagnosticMessage;
-use rustc_errors::{DiagnosticBuilder, Handler, IntoDiagnostic};
+use rustc_errors::{
+    DiagnosticArgValue, DiagnosticBuilder, Handler, IntoDiagnostic, IntoDiagnosticArg,
+};
 use rustc_hir as hir;
 use rustc_hir::def_id::DefId;
 use rustc_index::IndexVec;
@@ -262,6 +264,12 @@ impl<'tcx> fmt::Display for LayoutError<'tcx> {
             LayoutError::Cycle => write!(f, "a cycle occurred during layout computation"),
             LayoutError::ReferencesError(_) => write!(f, "the type has an unknown layout"),
         }
+    }
+}
+
+impl<'tcx> IntoDiagnosticArg for LayoutError<'tcx> {
+    fn into_diagnostic_arg(self) -> DiagnosticArgValue<'static> {
+        self.to_string().into_diagnostic_arg()
     }
 }
 

--- a/tests/ui/layout/layout-cycle.rs
+++ b/tests/ui/layout/layout-cycle.rs
@@ -1,0 +1,31 @@
+// build-fail
+//~^ ERROR: a cycle occurred during layout computation
+//~| ERROR: cycle detected when computing layout of
+
+// Issue #111176 -- ensure that we do not emit ICE on layout cycles
+
+use std::mem;
+
+pub struct S<T: Tr> {
+    pub f: <T as Tr>::I,
+}
+
+pub trait Tr {
+    type I: Tr;
+}
+
+impl<T: Tr> Tr for S<T> {
+    type I = S<S<T>>;
+}
+
+impl Tr for () {
+    type I = ();
+}
+
+fn foo<T: Tr>() -> usize {
+    mem::size_of::<S<T>>()
+}
+
+fn main() {
+    println!("{}", foo::<S<()>>());
+}

--- a/tests/ui/layout/layout-cycle.stderr
+++ b/tests/ui/layout/layout-cycle.stderr
@@ -1,0 +1,11 @@
+error[E0391]: cycle detected when computing layout of `S<S<()>>`
+   |
+   = note: ...which requires computing layout of `<S<()> as Tr>::I`...
+   = note: ...which again requires computing layout of `S<S<()>>`, completing the cycle
+   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
+
+error: failed to get layout for S<S<()>>: a cycle occurred during layout computation
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0391`.

